src/server/graceful.rs

```diff
@@ -0,0 +1,620 @@
+            == 1
+        {
+            let mut waker_list = self.waker_list.lock().unwrap();
+            for (_, waker) in waker_list.iter_mut() {
+                if let Some(waker) = waker.take() {
+                    waker.wake();
+                }
+            }
+        }
+    }
+}
+
+pin_project! {
+    /// A wrapper around a future that's being watched for graceful shutdown.
+    ///
+    /// This is returned by [`GracefulShutdown::watch`].
+    ///
+    /// # Panics
+    ///
+    /// This future might panic if it is polled
+    /// after the internal future has already returned `Poll::Ready` before.
+    ///
+    /// Whether or not this future panics in such cases is an implementation detail
+    /// and should not be relied upon.
+    pub struct GracefulFuture<C> {
+        #[pin]
+        future: GracefulConnectionFuture<C, GracefulWaiter>,
+        state: Option<Arc<GracefulState>>,
+    }
+}
+
+impl<C> Debug for GracefulFuture<C> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("GracefulFuture").finish()
+    }
+}
+
+impl<C> Future for GracefulFuture<C>
+where
+    C: GracefulConnection,
+{
+    type Output = C::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        match this.future.poll(cx) {
+            Poll::Ready(v) => {
+                this.state.take().unwrap().unsubscribe();
+                Poll::Ready(v)
+            }
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+/// A future that blocks until a graceful shutdown is complete.
+pub struct GracefulWaiter {
+    state: GracefulWaiterState,
+}
+
+impl GracefulWaiter {
+    fn new(state: Arc<GracefulState>) -> Self {
+        Self {
+            state: GracefulWaiterState { state, key: None },
+        }
+    }
+}
+
+impl Future for GracefulWaiter {
+    type Output = ();
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Self::Output> {
+        let state = &mut self.state;
+
+        if state
+            .state
+            .counter
+            .load(std::sync::atomic::Ordering::SeqCst)
+            == 0
+        {
+            return Poll::Ready(());
+        }
+
+        let mut waker_list = state.state.waker_list.lock().unwrap();
+
+        if state
+            .state
+            .counter
+            .load(std::sync::atomic::Ordering::SeqCst)
+            == 0
+        {
+            // check again in case of race condition
+            return Poll::Ready(());
+        }
+
+        let waker = Some(cx.waker().clone());
+        state.key = Some(match state.key.take() {
+            Some(key) => {
+                *waker_list.get_mut(key).unwrap() = waker;
+                key
+            }
+            None => waker_list.insert(waker),
+        });
+
+        Poll::Pending
+    }
+}
+
+struct GracefulWaiterState {
+    state: Arc<GracefulState>,
+    key: Option<usize>,
```

Comment:
Does this `key` describe the cursor of the state?

Reply:
The key is set in case it already has a waker registered for it. If so then we replace it instead of generating a new one in case it got polled again before woken up